src/core/queriesObserver.ts

```diff
@@ -64,78 +64,116 @@ export class QueriesObserver extends Subscribable<QueriesObserverListener> {
   }
 
   getOptimisticResult(queries: QueryObserverOptions[]): QueryObserverResult[] {
-    return queries.map((options, index) => {
-      const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-      return this.getObserver(defaultedOptions, index).getOptimisticResult(
-        defaultedOptions
-      )
+    return this.getUpdatedObservers(queries).newResult
+  }
+
+  private getUpdatedObservers(
+    queries: QueryObserverOptions[],
+    notifyOptions?: NotifyOptions
+  ): QueryObserverUpdate {
+    const prevObservers = this.observers
+    const prevObserversMap = this.observersMap
+    const newResult: QueryObserverResult[] = []
+    const newObservers: QueryObserver[] = []
+    const newObserversMap: Record<string, QueryObserver> = {}
+
+    const defaultedQueryOptions = queries.map(options =>
+      this.client.defaultQueryObserverOptions(options)
+    )
+    const matchingObservers = defaultedQueryOptions
+      .map(options => {
+        const match = prevObserversMap[options.queryHash!]
+        if (match != null) {
+          match.setOptions(options, notifyOptions)
+          return match
+        }
+        return null
+      })
+      .filter(notNullOrUndefined)
+
+    const matchedQueryHashes = matchingObservers.map(
+      observer => observer?.options.queryHash
+    )
+    const unmatchedQueries = defaultedQueryOptions.filter(
+      options => !matchedQueryHashes.includes(options.queryHash)
+    )
+
+    const unmatchedObservers = prevObservers.filter(
+      prevObserver => !matchingObservers.includes(prevObserver)
+    )
+
+    const newlyMatchedOrCreatedObservers = unmatchedQueries.map(options => {
+      if (options.keepPreviousData && unmatchedObservers.length > 0) {
+        // use the first observer but no longer matched query to keep query data for any new queries
+        const firstObserver = unmatchedObservers.splice(0, 1)[0]
+        if (firstObserver !== undefined) {
+          firstObserver.setOptions(options, notifyOptions)
```

Comment:
you mentioned in the PR description:

> This does cause some extra updates, as can be seen in the tests. If I understand correctly, they probably are due to the setOptions calls being called both on getOptimisticResult and updateObservers calls.

I don't think that we _can_ call `setOptions` in `getOptimisticResult`, because `getOptimisticResult` is called _during_ render. It doesn't have any side-effects in the current version, where `updateObserver` is only called:
- in the constructor
- in an effect in `useQueries`
which are both legal by the rules of react.

`setOptions` potentially triggers a fetch, so I'm afraid we really can't do that 🚨

Reply:
Ah, you are right! Thank you, I must have been too busy trying to see if I could call `setOptions` without triggering a re-render to stop to think if I should 😅 

I now refactored the code so the search part is all encapsulated to `findMatchingObservers`. Then `updateObservers` will separately call `setOptions` on the returned list. 

This gets rid of the extra renders, but there still seems to be some small change in the order of the updates compared to master. This means I couldn't completely revert the tests in `useQueries.test.tsx` back. However, the difference seems to rather small: two renders back to back, which previously both showed the query with keepPreviousData as fetching and `isPreviousData:true`, whereas now the first round still shows it not fetching yet (`isFetching: false`). Couldn't quite track why this happens, but it doesn't seem to have any visible effect, so I just did the small change to the tests.

Reply:
that's fine I think 👍